### PR TITLE
improve cat design / performance

### PR DESCRIPTION
**Pull request by vtjnash**
This used to make a lot of references to design issues with the SparseArrays package (https://github.com/JuliaLang/julia/issues/2326 / https://github.com/JuliaLang/julia/pull/20815), which result in a non-sensical dispatch arrangement, and contribute to a slow loading experience do to the illogical Unions that must be checked by subtyping.

Requires similar those issues to be fixed in SparseArrays first (https://github.com/JuliaSparse/SparseArrays.jl/pull/384) before merging this.

It is hard to get a reliable measure of the exact impact, since that measurement fluctuates a bit between builds due to other factors. But we can see this uses a bit less memory now, and I had instrumented it previously to measure that this cost 0.5s of load time, and that cost went do pretty much to zero after this change.
```
julia> @time using OmniPackage
 26.654442 seconds (20.52 M allocations: 1.322 GiB, 9.15% gc time, 29.93% compilation time: 86% of which was recompilation) # before
 26.697070 seconds (20.38 M allocations: 1.322 GiB, 8.54% gc time, 27.20% compilation time: 85% of which was recompilation) # after
```

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1982,19 +1982,18 @@ julia> cat(1, [2], [3;;]; dims=Val(2))
 @inline cat(A...; dims) = _cat(dims, A...)
 # `@constprop :aggressive` allows `catdims` to be propagated as constant improving return type inference
 @constprop :aggressive _cat(catdims, A::AbstractArray{T}...) where {T} = _cat_t(catdims, T, A...)
+const ArrayConcatGroup = Union{Number, AbstractArray}
```
Comment (JeffBezanson) [inline base/abstractarray.jl:1985]: Is this used anywhere?

**Review by JeffBezanson — APPROVED** [review]

**Comment by vtjnash** [comment]
Seems like SparseArrays has been broken on main for awhile, so we are waiting for this to be fixed before we can move ahead with this PR: https://github.com/JuliaSparse/SparseArrays.jl/pull/363#issuecomment-1479980510. It recently switched to using the JLL, but seems to still simultaneously be vendering its own copy of the JLL which is broken.

**Comment by Wimmerer** [comment]
I'm going to be trying to fix the SuiteSparse issue this weekend, I'm not sure when Tim Davis would release the update with my fix but he's pretty responsive.

**Comment by ViralBShah** [comment]
https://github.com/JuliaLang/julia/pull/48977 is ready now, which will pull in the updated SuiteSparse and the SparseArrays hvcat - but that PR doesn't pass CI.

**Comment by ViralBShah** [comment]
@vtjnash Is this ok to backport to 1.10?

**Comment by vtjnash** [comment]
Sure. It was ready before that, but just was waiting on SparseArrays. The performance boost to loading and running seemed worth it for backport.

**Comment by vtjnash** [comment]
It is expected, though is type-piracy by SparseArrays, so is an issue for that repo, not here:
 https://github.com/JuliaSparse/SparseArrays.jl/blob/0eb9c0472de0a392afb0bb70aef1cf68ae5bcb2d/src/sparsevector.jl#L1178-L1181
https://github.com/JuliaSparse/SparseArrays.jl/blob/0eb9c0472de0a392afb0bb70aef1cf68ae5bcb2d/src/abstractsparse.jl#L76

**Pull request merged**